### ISSUE TITLE
Question regarding kernel size and stride direction in the MRD module

### ISSUE THREAD

**Issue by jaywalnut310** [descriptinc/descript-audio-codec#36]
First and foremost, I'd like to thank for your great work. It's truly impressive!

By the way, I have a question about the MRD module. When I looked into the implementation, it appears that the larger kernel size and stride are applied along the frequency axis of the spectrogram. However, I believe it might be more appropriate to apply them along the time axis. If it's the case, was this done intentionally?

```python
# https://github.com/descriptinc/descript-audio-codec/blob/c7cfc5d2647e26471dc394f95846a0830e7bec34/dac/model/discriminator.py#L101
class MRD(nn.Module):
    def __init__(
        ...
        convs = lambda: nn.ModuleList(
            [
                ...
                WNConv2d(ch, ch, (3, 9), (1, 2), padding=(1, 4)),
                ...
            ]
        )
        self.band_convs = nn.ModuleList([convs() for _ in range(len(self.bands))])
        ...

    def spectrogram(self, x):
        ...
        x = rearrange(x, "b 1 f t c -> (b 1) c t f")
        # Split into bands
        x_bands = [x[..., b[0] : b[1]] for b in self.bands]
        return x_bands

    def forward(self, x):
        x_bands = self.spectrogram(x)
        ...
        for band, stack in zip(x_bands, self.band_convs):
            for layer in stack:
                band = layer(band)
                ...
        ...
```

**Comment by ritheshkumar95** [comment]
Thanks for bringing this to our notice. This was an unintentional error when porting the code from the original implementation of Multi-resolution Spectrogram Discriminator (MRSD) from UnivNet. You're right about noting that the larger kernel sizes + striding should be applied on the time axis and not the frequency axis.

Although we don't expect the results to change much (from past experience with these things), we will soon launch a few training runs to cross-check.

We will also be updating the code in this repo to fix the error, because we believe this updated recipe for audio generation is widely applicable beyond audio codec learning and we wouldn't want a mistake to continue existing 😅

**Comment by jaywalnut310** [comment]
Thank you for the prompt response and acknowledgement! It's great to hear that the issue will be addressed soon👍

**Comment by BridgetteSong** [comment]
@ritheshkumar95 @jaywalnut310 In the paper [High Fidelity Neural Audio Compression](https://arxiv.org/pdf/2210.13438.pdf) **3.4 Chapter**:
_`Each sub-network is composed of a 2D convolutional layer (using kernel size 3 x 8 with 32 channels), followed by 2D convolutions with increasing dilation rates in the time dimension of 1, 2 and 4, and a stride of 2 over the frequency axis. `_
Maybe it is a bit different from **UnivNet**, although it may not matter.

**Comment by lin7275** [comment]
> UnivNet

What should the fix  look like? I try to rearange the t and f. But does not work.

**Comment by pseeth** [comment]
This should be the fix:

```python
class MRD(nn.Module):
    def __init__(
        self,
        window_length: int,
        hop_factor: float = 0.25,
        sample_rate: int = 44100,
        bands: list = BANDS,
    ):
        """Complex multi-band spectrogram discriminator.
        Parameters
        ----------
        window_length : int
            Window length of STFT.
        hop_factor : float, optional
            Hop factor of the STFT, defaults to ``0.25 * window_length``.
        sample_rate : int, optional
            Sampling rate of audio in Hz, by default 44100
        bands : list, optional
            Bands to run discriminator over.
        """
        super().__init__()

        self.window_length = window_length
        self.hop_factor = hop_factor
        self.sample_rate = sample_rate
        self.stft_params = STFTParams(
            window_length=window_length,
            hop_length=int(window_length * hop_factor),
            match_stride=True,
        )

        n_fft = window_length // 2 + 1
        bands = [(int(b[0] * n_fft), int(b[1] * n_fft)) for b in bands]
        self.bands = bands

        ch = 32
        convs = lambda: nn.ModuleList(
            [
                WNConv2d(2, ch, (3, 9), (1, 1), padding=(1, 4)),
                WNConv2d(ch, ch, (3, 9), (1, 2), padding=(1, 4)),
                WNConv2d(ch, ch, (3, 9), (1, 2), padding=(1, 4)),
                WNConv2d(ch, ch, (3, 9), (1, 2), padding=(1, 4)),
                WNConv2d(ch, ch, (3, 3), (1, 1), padding=(1, 1)),
            ]
        )
        self.band_convs = nn.ModuleList([convs() for _ in range(len(self.bands))])
        self.conv_post = WNConv2d(ch, 1, (3, 3), (1, 1), padding=(1, 1), act=False)

    def spectrogram(self, x):
        x = AudioSignal(x, self.sample_rate, stft_params=self.stft_params)
        x = torch.view_as_real(x.stft())
        x = rearrange(x, "b 1 f t c -> (b 1) c f t")
        # Split into bands
        x_bands = [x[..., b[0] : b[1], :] for b in self.bands]
        return x_bands

    def forward(self, x):
        x_bands = self.spectrogram(x)
        fmap = []

        x = []
        for band, stack in zip(x_bands, self.band_convs):
            for layer in stack:
                band = layer(band)
                fmap.append(band)
            x.append(band)

        x = torch.cat(x, dim=-2)
        x = self.conv_post(x)
        fmap.append(x)

        return fmap
```

Main changes are 

```python
x = rearrange(x, "b 1 f t c -> (b 1) c f t") # t f became f t in the after
# Split into bands
x_bands = [x[..., b[0] : b[1], :] for b in self.bands] # frequency is now 2nd to last dim
```

and further down, concatenate on the frequency axis

```python
x = torch.cat(x, dim=-2)
```

I'll try to run this soon and compare!
